diff -aur stunnel-5.43/src/client.c stunnel-5.43-platform9/src/client.c
--- stunnel-5.46/src/client.c	2017-11-03 02:10:03.000000000 -0700
+++ stunnel-5.43-platform9/src/client.c	2018-06-05 16:01:04.992978854 -0700
@@ -772,6 +772,7 @@
                 c->ssl_ptr-=(size_t)num;
                 memset(c->ssl_buff+c->ssl_ptr, 0, (size_t)num); /* paranoia */
                 c->sock_bytes+=(size_t)num;
+                s_log(LOG_NOTICE, "wrote %ld byte(s) to socket", num);
                 watchdog=0; /* reset the watchdog */
             }
         }
@@ -792,6 +793,7 @@
                 break; /* do not reset the watchdog */
             default:
                 c->sock_ptr+=(size_t)num;
+                s_log(LOG_NOTICE, "read %ld byte(s) from socket", num);
                 watchdog=0; /* reset the watchdog */
             }
         }
@@ -820,6 +822,7 @@
                 c->sock_ptr-=(size_t)num;
                 memset(c->sock_buff+c->sock_ptr, 0, (size_t)num); /* paranoia */
                 c->ssl_bytes+=(size_t)num;
+                s_log(LOG_NOTICE, "wrote %ld byte(s) to TLS", num);
                 watchdog=0; /* reset the watchdog */
                 break;
             case SSL_ERROR_WANT_WRITE: /* buffered data? */
@@ -877,6 +880,7 @@
                     break; /* do not reset the watchdog */
                 }
                 c->ssl_ptr+=(size_t)num;
+                s_log(LOG_NOTICE, "read %ld byte(s) from TLS", num);
                 watchdog=0; /* reset the watchdog */
                 break;
             case SSL_ERROR_WANT_WRITE:
